### PR TITLE
10438 respect closed books date

### PR DESCRIPTION
**Pull request by DevneyHamilton**

### CODE REVIEW

File: app/models/accounting/interest_calculator.rb [hunk 1]
```diff
@@ -205,10 +234,20 @@ def month_boundaries(d1, d2)
     end
 
     def add_int_tx?(txs, prev_tx)
-      return true if txs.nil?
-      if txs
-       return true if prev_tx && prev_tx.principal_balance > 0 && txs.none?(&:interest?)
+      if txs.nil? # this is an end of month day with no txns
```
Comment (DevneyHamilton) [inline app/models/accounting/interest_calculator.rb:236]: This method seems to have too much nesting. I think we will want to know the difference between "no end of month int txn before closed books date" vs "there was a txn that wasn't paired with an interest txn before closed books date." I think the "no end of month int txn" could get really annoying and we could remove it if so; need to try it out with real data and see how common it is.

**Comment by DevneyHamilton** [comment]
@meltheadorable Question about testing here. The interest calculator spec has an elaborate set up (probably necessary) and elaborate expectations. The bluntest way to test closed book dates would be to copy all those existing specs, set a closed books date after those txns dates, and they are not updated. Should I try to wrangle the existing specs into a 'shared example' approach, and run them once with a closed book date before, and once with a closed book date after the txns' dates? or should I take this oppty to try to simplify the specs overall? or something else?

**Comment by DevneyHamilton** [comment]
@meltheadorable also could you please review the changes so far and let me know if they seem to be on track? 

resolved - this is ready for full review now

**Comment by DevneyHamilton** [comment]
> @meltheadorable Question about testing here. The interest calculator spec has an elaborate set up (probably necessary) and elaborate expectations. The bluntest way to test closed book dates would be to copy all those existing specs, set a closed books date after those txns dates, and they are not updated. Should I try to wrangle the existing specs into a 'shared example' approach, and run them once with a closed book date before, and once with a closed book date after the txns' dates? or should I take this oppty to try to simplify the specs overall? or something else?

I think I have this figured out.

**Comment by DevneyHamilton** [comment]
> @meltheadorable Question about testing here. The interest calculator spec has an elaborate set up (probably necessary) and elaborate expectations. The bluntest way to test closed book dates would be to copy all those existing specs, set a closed books date after those txns dates, and they are not updated. Should I try to wrangle the existing specs into a 'shared example' approach, and run them once with a closed book date before, and once with a closed book date after the txns' dates? or should I take this oppty to try to simplify the specs overall? or something else?

resolved - I added specs

**Comment by DevneyHamilton** [comment]
For QA:  I should try this with staging data on local first - it could blow up with a lot of errors showing for loan update etc.

**Review by meltheadorable — APPROVED** [review]
This looks great

**Comment by emitche** [comment]
image

**Comment by emitche** [comment]
Adding a closed book date and removing it results in an error on the update all loans task.

image

`if prev_tx.txn_date > @closed_books_date`

Update: This is a valid bug that will need to be addressed. This can be addressed outside of this pull request.

**Comment by emitche** [comment]
Update: Resolved

After adding back in a closed book date, I get: 

`unknown attribute 'custom_data' for Accounting::ProblemLoanTransaction.`

related to this line:

`          ::Accounting::ProblemLoanTransaction.create(loan: loan, accounting_transaction: prev_tx, message: :attempted_new_int_txn_before_closed_books_date, custom_data: {cbd: @closed_books_date, txn_date: prev_tx.txn_date}, level: :warning)
`

**Pull request merged**